user_news/forms.py

```diff
@@ -21,3 +22,8 @@ def clean_send_email_notification(self):
         send = self.cleaned_data.get('send_email_notification')
         self.instance.send_email_notification = send
         return send
+
+    def clean_send_email_reminder(self):
+        send = self.cleaned_data.get('send_email_reminder')
```

Comment:
Just curious, do you know why we do this/what it does?

Reply:
This is a good question and it took me awhile to figure it out. From what I can tell, this allows us to override needing to store the send_email_notification and send_email_reminder in the database (hence no migration needed). We have the email notification set to check and send an email whenever the content of an Outage form are set to be saved to the database. The check boxes are added to the form, but they are not model fields and therefore will not be stored in the db. Instead, we override django's cleaned data functionality to add the contents of the form to the model instance so that we can determine if we send an email without then having to store it in the database.

Reply:
Ah, nice explanation, thanks! OK, this makes sense. I guess there is not much value in saving the flag in the DB. I initially thought that maybe we would store something like, the date the outage reminder was sent, and if it's NULL then it knows to send the reminder when periodically checking if any reminders should be sent. But, since we are scheduling the emails to occur at a precise time, there isn't much need to store it.

This actually just occurred to me while writing this... because we are scheduling the emails at a future dates, I think we might have an issue where the emails get sent at the wrong time if the outage date is ever changed. Is there a way to clear out a scheduled task? Maybe we can do something where, if the outage is ever updated, the scheduled task is cleared and replaced with a new scheduled task?

Reply:
I'm wondering if we actually need to store the celery task ID in order to do this. Maybe we do need to store a field in the DB, like `notification_task_id`. I think the return of `apply_async` is something you can get a task ID from. Then, when updating an outage, you can revoke the task given its ID, create a new task, and store the new task ID. An optimization would be to not do any of this if the outage start date did not change.

Reply:
Looking into this a little deeper I think it can be done possibly one of three ways. The last two ways depend on if the signal is sent before or after the post is saved.

1) We assign the task_id and make it the same at the Outage model id. This way we can call the model id and tell celery to purge any tasks with that id before assigning a new task with that id. This would allow us to avoid migration.
2) If signal sent after post is saved. We update a new field that we exclude from the form that holds the email_reminder_task_id with the celery task id.
3) If signal sent before post is saved. We call the process in the send_email_reminder task in the model on save and apply the task id to the field.

Reply:
I took a bit of a hacky approach. I set the task id of the schedule_email task and attach a version number. This allows us to revoke the old task and create a new task with an incremented version number.